<a href="https://colab.research.google.com/github/xthomas/dataviz/blob/master/Klick_Role_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [17]:
import pandas as pd
import numpy as np
from google.colab import drive

In [18]:
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [19]:
!ls "/content/gdrive/My Drive/3. Klick/Corporate"

'Klick Role Analysis.ipynb'	   'Reporting Portal - User New Roles.xlsx'
 MaskingTest_MedicalStrategy.xlsx   UnmaskingTest_aFewRolesMore.xlsx


In [20]:
KlickRoles = pd.read_excel ('/content/gdrive/My Drive/3. Klick/Corporate/Reporting Portal - User New Roles.xlsx')

In [42]:
list (KlickRoles)

['UserID', 'Full Name', 'Title', 'RoleID', 'Labor Role', 'CST', 'Category']

In [43]:
KlickRoles.shape

(1587, 7)

In [22]:
#how many roles at each Client Service Team?
KlickRoles.groupby("CST")["RoleID"].count()

CST
Brave            19
Edge             83
Epic            103
Katalyst          5
Klick Health    387
Klick Inc       123
Nova            231
Quest            82
Rev             109
Sensei           77
Spark           182
Stack             1
Team Takeda     185
Name: RoleID, dtype: int64

In [23]:
KlickRoles.groupby("Labor Role").count()

UserID  Full Name  ...  RoleID  CST
Labor Role                                                        ...             
Activation Strategy                                 4          4  ...       4    4
Activation Strategy, Director                      15         15  ...      15   15
Activation Strategy, Executive Vice President       1          1  ...       1    1
Activation Strategy, Junior                         6          6  ...       6    6
Activation Strategy, Manager                        2          2  ...       2    2
...                                               ...        ...  ...     ...  ...
Web Development                                    66         66  ...      66   66
Web Development, Director                           1          1  ...       1    1
Web Development, Junior                             2          2  ...       2    2
Web Development, Manager                            1          1  ...       1    1
Web Development, Senior                            16         16  ...      16   16

[222 rows x 5 columns]

In [24]:
# Looks like the Role ID prefix can be used to categorize all these roles... 
KlickRoles.RoleID.unique()

array(['TECHMMGD', 'EXCTVMGD', 'USBLTVPR', 'TECHMDIR', 'WEBDVSNR',
       'TECHMVPR', 'FINCESVP', 'DESGNDIR', 'PRMGTDIR', 'TECHMSDR',
       'PRMGTSTD', 'EXCTVSTD', 'PRMGTVPR', 'USBLTDIR', 'SLSTGVPR',
       'TECHMSVP', 'MDSEMVPR', 'PRMGTSVP', 'DESGNMGR', 'FINCEMGR',
       'BUSDVSVP', 'CLSRVMGD', 'EXCTVSVP', 'BUSDVVPR', 'EDITRVPR',
       'CLSRVVPR', 'SYSADSTD', 'SYSADDIR', 'BUSDVSDR', 'TECHMADR',
       'CLSRVSTD', 'PRMGTMGR', 'QLTASVPR', 'CONSTMGD', 'CRMNGADR',
       'CLSRVSDR', 'QLTASMGR', 'CORPTADR', 'APPDVSNR', 'CLSRVSVP',
       'DESGNSTD', 'APPDVMGR', 'ARTDRMGR', 'FINCEVPR', 'EDITRSVP',
       'CLSRVDIR', 'MDRSHVPR', 'CRMNGVPR', 'STRGYEVP', 'SYSADSNR',
       'FINCEDIR', 'PRMGTSDR', 'CRMSTSDR', 'ARTDRSTD', 'EDITRDIR',
       'STRGYSVP', 'CPYWRMGR', 'APPDVSTD', 'CONSTDIR', 'STRGYVPR',
       'OPRTNSTD', 'CORPTSTD', 'BUSANSNR', 'MDSEOVPR', 'ANLTCMGR',
       'CORPTVPR', 'OPRTNDIR', 'CORPTMGR', 'CONSTMGR', 'CORPTDIR',
       'EDITRSTD', 'MDSTGDIR', 'EDITRMGR', 'SLSTGSTD', 'PRDCRS

In [26]:
# but there are a LOT of roles...and looks like some overlap
# for example "MD" isn't _just_ **Media**, it's _also_ **Medical Strategy**
startswithmd = KlickRoles[KlickRoles.RoleID.str.startswith('MD')]
startswithmd.RoleID.unique()

array(['MDSEMVPR', 'MDRSHVPR', 'MDSEOVPR', 'MDSTGDIR', 'MDPLNDIR',
       'MDSEOSTD', 'MDPRGVPR', 'MDTTVSDR', 'MDAOPVPR', 'MDPLNSVP',
       'MDSTGSVP', 'MDPLNVPR', 'MDSEMMGR', 'MDSTGMGD', 'MDSEMDIR',
       'MDSTGVPR', 'MDTTVSTD', 'MDAOPADR', 'MDPLNADR', 'MDSTGSDR',
       'MDPLNSDR', 'MDPLNMGR', 'MDPLNSTD', 'MDAOPMGR', 'MDSTGMGR',
       'MDPLNJNR', 'MDPRGMGR', 'MDPDSMGR', 'MDSTGADR', 'MDAOPJNR',
       'MDTTVADR', 'MDSEOADR', 'MDPDSSTD', 'MDPDSDIR', 'MDPRGADR',
       'MDSEMADR', 'MDRSHSTD', 'MDRSHJNR', 'MDPDSJNR', 'MDTTVJNR',
       'MDSEMSTD', 'MDPDSSDR', 'MDSTGSTD', 'MDPRGSTD', 'MDSEOMGR',
       'MDPDSADR', 'MDPRGJNR', 'MDPRGSDR', 'MDSTGJNR'], dtype=object)

In [ ]:
startswithmd.head(10)

UserID  ...           CST
25     4001  ...  Klick Health
65     4601  ...  Klick Health
107    5016  ...  Klick Health
125    5172  ...  Klick Health
138    5305  ...          Nova
177    5640  ...  Klick Health
189    5726  ...  Klick Health
232    5979  ...      Katalyst
238    6016  ...  Klick Health
259    6130  ...   Team Takeda

[10 rows x 6 columns]

In [ ]:
startswithmd.shape

(132, 6)

☝ 132...looks 'scannable' in a spreadsheet

Let's see if we can decode **Departments** by the **RoleID prefix**


First...create 'masks' for the main departments I want to compare

---




In [ ]:
# **MEDIA**
mask1 = ((startswithmd.RoleID.str.startswith('MD')) & (~startswithmd.RoleID.str.startswith('MDST')))

# **MEDICAL STRATEGY**
mask2 = startswithmd.RoleID.str.startswith('MDST')

In [ ]:
# then write a new CATEGORY column, based on what mask fits them
# let's start with something easier to see, like this "MD" slice
startswithmd['Category'] = np.select([mask1, mask2],['Media','Medical Strategy'],'Confusing')
startswithmd.to_excel('/content/gdrive/My Drive/3. Klick/Corporate/MaskingTest_MedicalStrategy.xlsx')

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


A [quick scan in Google Sheets](https://drive.google.com/file/d/1tb3GCNiiu4BeFIEg7jcrF2Q9Ur6Ubhju/view?usp=sharing) ☜ of **132 RoleIDs starting with "MD"** looks like those 'masks' work well enough!

Now to create more 'masks'—let's call them filters—for those other department prefixes, to explore the **department composition variance** from one CST to another

In [34]:
# **01. MEDIA**
f1 = ((KlickRoles.RoleID.str.startswith('MD')) & (~startswithmd.RoleID.str.startswith('MDST')))

# **02. REGULATORY COMPLIANCE**
f2 = KlickRoles.RoleID.str.startswith('EDIT')


# **03. PROJECT MANAGEMENT**
f3 = KlickRoles.RoleID.str.startswith('PRMG')

# **04. TECHNICAL SERVICES**
f4 = (KlickRoles.RoleID.str.startswith('TECH') | 
         KlickRoles.RoleID.str.startswith('APP') | 
         KlickRoles.RoleID.str.startswith('QLTA') | 
         KlickRoles.RoleID.str.startswith('WEBDV') |
         KlickRoles.RoleID.str.startswith('SYSAD') | 
         KlickRoles.RoleID.str.startswith('BUSAN') |                  
         KlickRoles.RoleID.str.startswith('WEBDV') |
         KlickRoles.RoleID.str.startswith('CRMOP') |
         ((KlickRoles.RoleID.str.startswith('CRM')) & (~startswithmd.RoleID.str.startswith('CRMST')))      
         )

# **05. ACCOUNT MANAGEMENT**
f5 = KlickRoles.RoleID.str.startswith('CLSR')

# **06. CORPORATE OVERHEAD**
f6 = (KlickRoles.RoleID.str.startswith('EXCT') | 
         KlickRoles.RoleID.str.startswith('FINCE') | 
         KlickRoles.RoleID.str.startswith('CORP') | 
         KlickRoles.RoleID.str.startswith('BUSDV') |
         KlickRoles.RoleID.str.startswith('OPRT')
         )

# **07. MEDICAL STRATEGY**
f7 = KlickRoles.RoleID.str.startswith('MDST') 

# **08. BRAND STRATEGY**
f8 = KlickRoles.RoleID.str.startswith('BRST')

# **09. ACTIVATION STRATEGY**
f9 = KlickRoles.RoleID.str.startswith('STRG')

# **10. UX STRATEGY**
f10 = (KlickRoles.RoleID.str.startswith('USBLT') | 
         KlickRoles.RoleID.str.startswith('EXPDS')
         )

# **11. CRM STRATEGY**
f11 = KlickRoles.RoleID.str.startswith('CRMST')

# **12. SOCIAL STRATEGY**
f12 = KlickRoles.RoleID.str.startswith('SLST')

# **13. ANALYTICS**
f13 = (KlickRoles.RoleID.str.startswith('ANLT') | 
         KlickRoles.RoleID.str.startswith('DATSC')
         )

# **14. CREATIVE**
f14 = (KlickRoles.RoleID.str.startswith('DESGN') | 
         KlickRoles.RoleID.str.startswith('CPYWR') | 
         KlickRoles.RoleID.str.startswith('CRMNG') | 
         KlickRoles.RoleID.str.startswith('ART') |
         KlickRoles.RoleID.str.startswith('VIDED') |
         KlickRoles.RoleID.str.startswith('PRD')
         )

# **15. FANCYPANTS GROUPS**
f15 = (KlickRoles.RoleID.str.startswith('CONST') | 
         KlickRoles.RoleID.str.startswith('SCIEN')
         )

In [35]:
# good enough, write those to a new CATEGORY column, based on what mask fits them
# dump it to a spreadsheet to scan for matches
KlickRoles['Category'] = np.select([f1,f2,f3,f4,f5,f6,f7,f8,f9,f10,f11,f12,f13,f14,f15],[
                                                   'Media',
                                                   'Regulatory Compliance',
                                                   'Project Management',
                                                   'Technical Services',
                                                   'Account Management',
                                                   'Corporate Overhead',
                                                   'Medical Strategy',
                                                   'Brand Strategy',
                                                   'Activation Strategy',
                                                   'UX Strategy',
                                                   'CRM Strategy',
                                                   'Social Strategy',
                                                   'Analytics',
                                                   'Creative',
                                                   'Fancy Disciplines'
                                                   ],'Confusing')
KlickRoles.to_excel('/content/gdrive/My Drive/3. Klick/Corporate/Klick Roles - Categorized.xlsx')

SOME INTERESTING VIEWS TO CONSIDER RENDERING

* Corporate Overhead (corporate vs rest)
* Creative Overhead (Creative vs rest)
* Strategy vs Project Management for each CST
* Strategy Composition by each CST



In [38]:
strategy = KlickRoles[KlickRoles.Category.str.endswith("Strategy")]
strategy.shape

(178, 7)

In [48]:
stratByCst = pd.pivot_table(strategy,index=["CST"],
               values=["RoleID"],
               columns=["Category"],
               aggfunc=[len],fill_value=0)
stratByCst.to_csv('/content/gdrive/My Drive/3. Klick/Corporate/Strategy Types, by CST.csv')